### PR TITLE
[DML] Add int4 QDQ

### PR DESCRIPTION
**Pull request by PatriceVignola**

### CODE REVIEW

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorElementWise.cpp [hunk 1]
```diff
@@ -592,12 +593,129 @@ class DmlOperatorElementwiseQLinear : public DmlOperator
...
+                quantizationOpDesc.QuantizationType = DML_QUANTIZATION_TYPE_SCALE_ZERO_POINT;
+                quantizationOpDesc.QuantizationTensorCount = static_cast<uint32_t>(quantizationTensors.size());
+                quantizationOpDesc.QuantizationTensors = quantizationTensors.data();
+                quantizationOpDesc.OutputTensor = &outputDescs[0];
+                DML_OPERATOR_DESC quantizationOpDmlDesc = { ApiTraits::OperatorDescTraits<TOperatorDesc>::Type, &quantizationOpDesc };
+
+                std::vector<const DML_OPERATOR_DESC*> opDescs;
+                opDescs.reserve(2);
+                opDescs.push_back(&zeroPointConstantDmlDesc);
+                opDescs.push_back(&quantizationOpDmlDesc);
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorElementWise.cpp:657]:
```suggestion
                std::array<const DML_OPERATOR_DESC*, 2> opDescs = {
                    &zeroPointConstantDmlDesc,
                    &quantizationOpDmlDesc
                };
```
Fixed size?

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorElementWise.cpp [hunk 2]
```diff
@@ -592,12 +593,129 @@ class DmlOperatorElementwiseQLinear : public DmlOperator
...
+            }
+            else
+            {
+                // For unsigned quantization, DML uses the midpoint of the datatype when zero point isn't provided (e.g. 8 for uint4 and 128 for uint8)
+                // since this is the most sane default in theory that represents the midpoint of the quantization. This is also the default used by various
+                // other frameworks and quantization tools. But because ONNX uses a default zero point of 0 no matter if it's signed or unsigned, we need
+                // to generate a constant zero point tensor with a value of 0 to override dml's default value when it isn't provided.
+                auto zeroPointDataType = std::is_same_v<TOperatorDesc, DML_QUANTIZE_OPERATOR_DESC> ? outputDataType : inputDataType;
+
+                // DML doesn't support int4 FILL_VALUE_CONSTANT yet, so simply create an int8 scalar and reinterpret it to an int4 tensor
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorElementWise.cpp:629]: > DML doesn't support int4 FILL_VALUE_CONSTANT yet

(resolveme) That's going to be interesting, especially with existing output content and strides... (but yes, I think int4 should be supported eventually for the most primitive operations like fill and identity, even if via a cast and bitwise mask/or in the worst case).

**Review by fdwr — APPROVED** [review]

File: onnxruntime/test/providers/cpu/tensor/quantize_linear_test.cc [hunk 3]
```diff
@@ -61,6 +74,19 @@ TEST(DequantizeLinearOpTest, UInt4) {
...
+  std::vector<int64_t> dims{5};
+  constexpr int unused_val = 0;
+
+  // Odd number of uint4 values to test packing/unpacking
+  test.AddInput<UInt4x2>("x", dims, {UInt4x2(0, 1), UInt4x2(3, 15), UInt4x2(2, unused_val)});
+  test.AddInput<float>("x_scale", {}, {2.0f});
+  test.AddOutput<float>("y", dims, {0.0f, 2.0f, 6.0f, 30.0f, 4.0f});
+  test.Run(OpTester::ExpectResult::kExpectSuccess, "", {kTensorrtExecutionProvider});
+}
+
```
[the remark this answers is not in the document]
Comment (PatriceVignola) [reply onnxruntime/test/providers/cpu/tensor/quantize_linear_test.cc:89]: That's what the `UInt4` test does (it uses a zero point of 1).

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorElementWise.cpp (hunk 1 above)
Comment (fdwr) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorElementWise.cpp]: Feedback not noticed?

File: onnxruntime/test/providers/cpu/tensor/quantize_linear_test.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (fdwr) [reply onnxruntime/test/providers/cpu/tensor/quantize_linear_test.cc:89]: Ah, an existing test outside the visible diff. 👍

**Review by fdwr — DISMISSED** [review]
👍

**Review by fdwr — APPROVED** [review]

**Pull request merged**